pkg/cloud/services/loadbalancer/loadbalancer.go

```diff
@@ -164,6 +87,144 @@ func (s *Service) ReconcileLoadBalancer(openStackCluster *infrav1.OpenStackClust
 	return nil
 }
 
+func (s *Service) getOrCreateLoadBalancer(openStackCluster *infrav1.OpenStackCluster, loadBalancerName, subnetID string) (*loadbalancers.LoadBalancer, error) {
+	lb, err := s.checkIfLbExists(loadBalancerName)
+	if err != nil {
+		return nil, err
+	}
+
+	if lb != nil {
+		return lb, nil
+	}
+
+	s.logger.Info(fmt.Sprintf("Creating load balancer in subnet: %q", subnetID), "name", loadBalancerName)
+
+	lbCreateOpts := loadbalancers.CreateOpts{
+		Name:        loadBalancerName,
+		VipSubnetID: subnetID,
+	}
+
+	lb, err = loadbalancers.Create(s.loadbalancerClient, lbCreateOpts).Extract()
+	if err != nil {
+		record.Warnf(openStackCluster, "FailedCreateLoadBalancer", "Failed to create load balancer %s: %v", loadBalancerName, err)
+		return nil, err
+	}
+
+	if err := s.waitForLoadBalancerActive(lb.ID); err != nil {
+		record.Warnf(openStackCluster, "FailedWaitForLoadBalancerActive", "Failed to wait for load balancer active %s with id %s: %v", loadBalancerName, lb.ID, err)
+		return nil, err
+	}
+
+	record.Eventf(openStackCluster, "SuccessfulCreateLoadBalancer", "Created load balancer %s with id %s", loadBalancerName, lb.ID)
+	return lb, nil
+}
+
+func (s *Service) getOrCreateListener(openStackCluster *infrav1.OpenStackCluster, listenerName, lbID string, port int) (*listeners.Listener, error) {
+	listener, err := s.checkIfListenerExists(listenerName)
+	if err != nil {
+		return nil, err
+	}
+
+	if listener != nil {
+		return listener, nil
+	}
+
+	s.logger.Info("Creating load balancer listener", "name", listenerName, "lb-id", lbID)
+
+	listenerCreateOpts := listeners.CreateOpts{
+		Name:           listenerName,
+		Protocol:       "TCP",
+		ProtocolPort:   port,
+		LoadbalancerID: lbID,
+	}
+	listener, err = listeners.Create(s.loadbalancerClient, listenerCreateOpts).Extract()
+	if err != nil {
+		record.Warnf(openStackCluster, "FailedCreateListener", "Failed to create listener %s: %v", listenerName, err)
+		return nil, err
+	}
+
+	if err := s.waitForLoadBalancerActive(lbID); err != nil {
+		record.Warnf(openStackCluster, "FailedWaitForLoadBalancerActive", "Failed to wait for load balancer active %s: %v", lbID, err)
```

Comment:
```suggestion
		record.Warnf(openStackCluster, "FailedCreateListener", "Failed to create listener %s: wait for load balancer active %s: %v", lbID, err)
```
I think I would prefer it like this. It's still possible to infer what failed (listener creation) and it can also be inferred in which step (wait for lb active) it failed

(would also apply to the other cases below)

Reply:
Thanks, I understand your intention. event name shows what. message shows why. I agree and will update.